plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange/ASTPreProcess.hs

```diff
@@ -0,0 +1,107 @@
+import qualified Data.Set                        as Set
+import           Development.IDE.GHC.Compat      (ContextInfo (MatchBind, TyDecl, ValBind),
+                                                  HieAST (..), Identifier,
+                                                  IdentifierDetails (identInfo),
+                                                  NodeInfo (NodeInfo, nodeIdentifiers),
+                                                  RefMap, Span, flattenAst,
+                                                  isAnnotationInNodeInfo,
+                                                  mkAstNode, mkRealSrcSpan,
+                                                  nodeInfoFromSource,
+                                                  realSrcSpanEnd,
+                                                  realSrcSpanStart)
+import           Development.IDE.GHC.Compat.Util (FastString)
+import           Prelude                         hiding (span)
+
+newtype PreProcessEnv a = PreProcessEnv
+    { preProcessEnvRefMap :: RefMap a
+    }
+
+-- | Make the AST more suitable for generating selection range.
+preProcessAST :: HieAST a -> Reader (PreProcessEnv a) (HieAST a)
+preProcessAST node = mergeImports node >>= mergeSignatureWithDefinition
+
+mergeImports :: HieAST a -> Reader (PreProcessEnv a) (HieAST a)
+mergeImports node = pure $ node { nodeChildren = children }
+  where
+    children = mapMaybe merge
+        . groupBy (\x y -> nodeIsImport x && nodeIsImport y)
+        . nodeChildren $ node
+
+    merge []  = Nothing
+    merge [x] = Just x
+    merge xs  = Just (createVirtualNode xs)
+
+nodeIsImport :: HieAST a -> Bool
+nodeIsImport = isAnnotationInAstNode ("ImportDecl", "ImportDecl")
+
+createVirtualNode :: [HieAST a] -> HieAST a
+createVirtualNode nodes = mkAstNode (NodeInfo mempty mempty mempty) span' nodes
+  where
+    span' = mkRealSrcSpan (minimum locations) (maximum locations)
+    locations = (\s -> [realSrcSpanStart s, realSrcSpanEnd s]) . nodeSpan =<< nodes
+
+mergeSignatureWithDefinition :: HieAST a -> Reader (PreProcessEnv a) (HieAST a)
+mergeSignatureWithDefinition node = do
+    refMap <- asks preProcessEnvRefMap
+    children' <- traverse mergeSignatureWithDefinition (nodeChildren node)
+    pure $ node { nodeChildren = reverse $ foldl' (go refMap) [] children' }
+  where
+    go :: RefMap a -> [HieAST a] -> HieAST a -> [HieAST a]
+    go _ [] node' = [node']
+    go refMap (prev:others) node' =
+        case mergeNearbySigDef refMap (prev, node') of
+            Nothing   -> node':prev:others
+            Just comb -> comb:others
+
+mergeNearbySigDef :: RefMap a -> (HieAST a, HieAST a) -> Maybe (HieAST a)
+mergeNearbySigDef refMap (n1, n2) = do
+    if not (("TypeSig", "Sig") `isAnnotationInAstNode` n1 && ("FunBind", "HsBindLR") `isAnnotationInAstNode` n2)
+    then Nothing
+    else do
+        typeSigId <- identifierForTypeSig n1
+        refs <- Map.lookup typeSigId refMap
```

Comment:
There's something clever going on here which is why you need this map, but I'm not sure what it is. Comment?

Reply:
Yes, indeed, this need a detailed explaination